contract/tests/test_contract.py

```diff
@@ -0,0 +1,107 @@
+        res = self.contract_line._onchange_product_id()
+        self.assertIn('uom_id', res['domain'])
+        self.contract_line.price_unit = 100.0
+
+        self.contract.partner_id = False
+        with self.assertRaises(ValidationError):
+            self.contract.recurring_create_invoice()
+        self.contract.partner_id = self.partner.id
+
+        self.contract.recurring_create_invoice()
+        self.invoice_monthly = self.env['account.invoice'].search(
+            [('contract_id', '=', self.contract.id)])
+        self.assertTrue(self.invoice_monthly)
+        new_date = self.current_date + relativedelta(
+            months=self.contract.recurring_interval)
+        self.assertEqual(self.contract.recurring_next_date,
+                         new_date.strftime('%Y-%m-%d'))
+
+        self.inv_line = self.invoice_monthly.invoice_line_ids[0]
+        self.assertAlmostEqual(self.inv_line.price_subtotal, 50.0)
+        self.assertTrue(self.inv_line.invoice_line_tax_ids)
+
+    def test_contract_daily(self):
+        self.contract_daily.recurring_create_invoice()
+        invoice_daily = self.env['account.invoice'].search(
+            [('contract_id', '=', self.contract_daily.id)])
+        self.assertTrue(invoice_daily)
+        new_date = self.current_date + relativedelta(
+            days=self.contract_daily.recurring_interval)
+        self.assertEqual(self.contract_daily.recurring_next_date,
+                         new_date.strftime('%Y-%m-%d'))
+
+    def test_contract_weekly(self):
+        self.contract_weekly.recurring_create_invoice()
+        invoices_weekly = self.env['account.invoice'].search(
+            [('contract_id', '=', self.contract_weekly.id)])
+        self.assertTrue(invoices_weekly)
+        new_date = self.current_date + relativedelta(
+            weeks=self.contract_weekly.recurring_interval)
+        self.assertEqual(self.contract_weekly.recurring_next_date,
+                         new_date.strftime('%Y-%m-%d'))
+
+    def test_onchange_partner_id(self):
+        self.contract.pricelist_id = False
+        self.contract._onchange_partner_id()
+        self.assertEqual(self.contract.pricelist_id,
+                         self.contract.partner_id.property_product_pricelist)
+
+    def test_onchange_recurring_invoices(self):
+        self.contract.recurring_next_date = False
+        self.contract._onchange_recurring_invoices()
+        self.assertEqual(self.contract.recurring_next_date,
+                         self.contract.date_start)
+
+    def test_check_journal(self):
+        contract_no_journal = self.contract.copy()
+        contract_no_journal.journal_id = False
+        journal = self.env['account.journal'].search([('type', '=', 'sale')])
+        journal.write({'type': 'general'})
+        with self.assertRaises(ValidationError):
+            contract_no_journal.recurring_create_invoice()
+        journal.write({'type': 'sale'})
```

Comment:
You don't need to write again the previous type, as all operations are rolled back when the method finishes, so no need to manually restore it.